File: tradedoublerandroid/src/main/java/com/tradedouble/tradedoublerandroid/network/NetClient.java
package com.tradedouble.tradedoublerandroid.network;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;


public class NetClient {

    public final OkHttpClient client;
    private static NetClient netClient;

    public NetClient() {
        client = initOkHttpClient();
    }

    private OkHttpClient initOkHttpClient() {
        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .readTimeout(10000, TimeUnit.MILLISECONDS)
                .connectTimeout(10000, TimeUnit.MILLISECONDS)
                .addInterceptor(loggingInterceptor)
                .build();
        return okHttpClient;
    }

    public static NetClient getNetClient() {
        if (netClient == null) {
            netClient = new NetClient();
        }
        return netClient;
    }


    public void callResponse(String url, ResultRequest resultRequest) throws IOException {
        Request request = new Request.Builder()
                .url(url)
                .get()
                .build();

        Call call = getNetClient().initOkHttpClient().newCall(request);
        call.enqueue(new Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                resultRequest.onFailure(-1);
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                if (response.isSuccessful()){
                    if (response.code() >= 200 && response.code() <=300) {
                        resultRequest.onResponseSuccess(response.code() );
                    }else{
                        resultRequest.onFailure(response.code());
                    }
                }
            }
        });

    }


}

File: tradedoublerandroid/src/main/java/com/tradedouble/tradedoublerandroid/network/NetClient.kt
package com.tradedouble.tradedoublerandroid.network

import okhttp3.*
import okhttp3.logging.HttpLoggingInterceptor
import java.io.IOException
import java.net.HttpRetryException
import java.util.concurrent.TimeUnit

class NetClient {
    val client: OkHttpClient
    private fun initOkHttpClient(): OkHttpClient {
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
        return OkHttpClient.Builder()
            .readTimeout(10000, TimeUnit.MILLISECONDS)
            .connectTimeout(10000, TimeUnit.MILLISECONDS)
            .addInterceptor(loggingInterceptor)
            .build()
    }

    @Throws(IOException::class)
    fun callResponse(url: String, resultRequest: ResultRequest) {

        val request = Request.Builder()
            .url(url)
            .get()
            .build()
        val call =
            netClient
                ?.initOkHttpClient()?.newCall(request)
        call?.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                resultRequest.onFailure( call.execute().code, e.message)
            }

            @Throws(IOException::class)
            override fun onResponse(
                call: Call,
                response: Response
            ) {
                if (response.isSuccessful) {
                    resultRequest.onResponseSuccess(
                        response.code,
                        responseBody = response.body.toString()
                    )
                }
            }
        })
    }

    companion object {
        var netClient: NetClient? = null
            get() {
                if (field == null) {
                    field =
                        NetClient()
                }
                return field
            }
            private set
    }

    init {
        client = initOkHttpClient()
    }
}